### PR TITLE
feat: support for multi channel licenses

### PR DESCRIPTION
**Pull request by pandemicsyn**
#### What this PR does / why we need it:

Add's support for multi channel licenses. 

Spec & behaviors

- kots install:
  - Prevalidates that the requested channel is present in the latest license (if online)
  - If channel slug does not match a channel in the license channel list we fail
  - We use the channel slug to determine what channel is being install. If channel slug is missing, default will be stable.
  - The installed channel will be recorded in the `app_version.selected_channel_id` , we do this by pulling from a new `requested_channel_slug` configmap entry during install and resolving that to a channel id
-kots upgrades:
  - We use app.selected_channel_id field to check for upgrades in those channels by passing that on to the replicated api.
- kots pull:
  - same basic behaviors as install for pre-validation
- airgap installs:
  - inherits pre validation from `kots install` minus the the attempt to get the latest license.
  - If channel slug does not match a channel in the license channel list we fail
- changing channels
  - online installs don't support changing channel in this iteration
  - airgap does as long as the channel is allowed by the license
- some kots related replicated api calls now include a query param of `selectedChannelId` to actually pass on what the selected channel is    

#### Which issue(s) this PR fixes:

Fixes https://app.shortcut.com/replicated/story/108865/multi-channel-license-update-kots-to-support-multi-channels

## Steps to reproduce

Online install early checks:

```
# starting config only access to Stable channel:

➜ replicated customer ls
ID                             NAME             CHANNELS        EXPIRES    TYPE    CUSTOM_ID
2jbffGP8wHSWH1l2mA7dM6ZeloV    multitest         Stable         Never      dev     Not Set
➜ bin/kots install --kotsadm-registry=ttl.sh/pandemicsyn --kotsadm-tag 24h floriandev/stable --license-file ~/Downloads/multichan.yaml
➜ replicated customer download-license --customer=2jbffGP8wHSWH1l2mA7dM6ZeloV -o ~/Downloads/multichan.yaml
```

Simple case where you ask to install a channel not in the licenses fails:

```
➜ bin/kots install --kotsadm-registry=ttl.sh/pandemicsyn --kotsadm-tag 24h floriandev/beta --license-file ~/Downloads/multichan.yaml
  • Validating registry information ✓
  • Checking for license update ✓
Error: requested channel not found in latest license
```

If you pass in an outdated license it will declined if the requested channel is not in the latest license:

```
# remove access to `Stable`
➜ replicated customer update --customer=2jbffGP8wHSWH1l2mA7dM6ZeloV --channel Beta --channel Unstable --default-channel Unstable --name multitest
ID                             NAME         CHANNELS          EXPIRES    TYPE    CUSTOM_ID
2jbffGP8wHSWH1l2mA7dM6ZeloV    multitest     Beta Unstable    Never      dev     Not Set
➜ bin/kots install --kotsadm-registry=ttl.sh/pandemicsyn --kotsadm-tag 24h floriandev/stable --license-file ~/Downloads/multichan.yaml
  • Validating registry information ✓
  • Checking for license update ✓
Error: requested channel not found in latest license
```

Install proceeds if your requested channel is present (even when its not the default channel):

```
# now installing with the Beta slug which should succeed
➜ replicated customer download-license --customer=2jbffGP8wHSWH1l2mA7dM6ZeloV -o ~/Downloads/multichan.yaml
➜ yq '.spec.channels' < ~/Downloads/multichan.yaml
- channelID: 2i9fCY11QaVowQWN1O32XFhUIlE
  channelName: Beta
  channelSlug: beta
  endpoint: https://replicated-app-pandemicsyn.okteto.repldev.com
  isSemverRequired: true
- channelID: 2i9fCYatEJfYChzAsL4bClT6E04
  channelName: Unstable
  channelSlug: unstable
  endpoint: https://replicated-app-pandemicsyn.okteto.repldev.com
  isDefault: true
➜ bin/kots install --kotsadm-registry=ttl.sh/pandemicsyn --kotsadm-tag 24h floriandev/beta --license-file ~/Downloads/multichan.yaml
  • Validating registry information ✓
  • Checking for license update ✓
Enter the namespace to deploy to: floriandev
  • Deploying Admin Console
    • Creating namespace ✓
    • Waiting for datastore to be ready ✓
```

Same pattern for airgap:

```
➜ yq '.spec.channels' < ~/Downloads/multitest/multitest.yaml
- channelID: 2i9fCY11QaVowQWN1O32XFhUIlE
  channelName: Beta
  channelSlug: beta
  endpoint: https://replicated-app-pandemicsyn.okteto.repldev.com
  isSemverRequired: true
- channelID: 2i9fCYatEJfYChzAsL4bClT6E04
  channelName: Unstable
  channelSlug: unstable
  endpoint: https://replicated-app-pandemicsyn.okteto.repldev.com
  isDefault: true
➜ bin/kots install --kotsadm-registry=ttl.sh/pandemicsyn --kotsadm-tag 24h floriandev/stable --license-file ~/Downloads/multitest/multitest.yaml --airgap-bundle ~/Downloads/multitest/floriandev-0.1.5.airgap
  • Validating registry information ✓
Error: requested channel not found in supplied license
➜ bin/kots install --kotsadm-registry=ttl.sh/pandemicsyn --kotsadm-tag 24h floriandev/beta --license-file ~/Downloads/multitest/multitest.yaml --airgap-bundle ~/Downloads/multitest/floriandev-0.1.5.airgap
  • Validating registry information ✓
  • Checking for license update ✓
Enter the namespace to deploy to: floriandev
  • Deploying Admin Console
    • Creating namespace ✓
    • Waiting for datastore to be ready ✓
Enter a new password for the admin console (6+ characters): ••••••••
  • Waiting for Admin Console to be ready ✓
  • Uploading app archive
  • Waiting for installation to complete ✓

  • Press Ctrl+C to exit
  • Go to http://localhost:8800 to access the Admin Console
```

With an airgap update via the ui:

Screenshot 2024-07-23 at 9 43 45 AM

Up to date Loom: https://www.loom.com/share/c82e9174906b427196a03a5d16ac3402?sid=7989633f-4ad7-4b86-9a5a-9c28d3ca8705

#### Does this PR introduce a user-facing change?

```release-note

- Add's support for at install time license update checks for online installs
- Add's support for multi-channel licenses 

```

### CODE REVIEW

File: cmd/kots/cli/install.go [hunk 1]
```diff
@@ -162,6 +162,16 @@ func InstallCmd() *cobra.Command {
...
 
 			upstream := pull.RewriteUpstream(args[0])
+			preferredChannelSlug, err := extractPreferredChannelSlug(upstream)
+			if err != nil {
+				return errors.Wrap(err, "failed to extract preferred channel slug")
+			}
+
+			// If we are passed a multi-channel license, verify that the requested channel is in the license
+			// so that we can warn the user immediately if it is not.
+			if license != nil && !slugInLicenseChannels(preferredChannelSlug, license) {
```
Comment (divolgin) [inline cmd/kots/cli/install.go:172]: we should get the latest license first if this is an online install

File: pkg/kotsutil/kots.go [hunk 2]
```diff
@@ -1598,3 +1600,62 @@ func GetECVersionFromAirgapBundle(airgapBundle string) (string, error) {
...
+		logger.Debug("requested channel slug not found in configmap, using top level channel id from license")
+		matchedChannelID = license.Spec.ChannelID
+	}
+	return matchedChannelID, nil
+}
+
+func FindChannelInLicense(channelID string, license *kotsv1beta1.License) (*kotsv1beta1.Channel, error) {
+	if channelID == "" || len(license.Spec.Channels) == 0 {
+		if license.Spec.ChannelID != channelID {
+			return nil, errors.New("channel not found in license")
```
Comment (divolgin) [inline pkg/kotsutil/kots.go:1632]: nit: same error in two places in the same function. it will be impossible to know which code path failed.

File: pkg/update/update.go [hunk 3]
```diff
@@ -29,7 +29,23 @@ func InitAvailableUpdatesDir() error {
...
-	updateCursor, err := kotsStore.GetCurrentUpdateCursor(app.ID, license.Spec.ChannelID)
+	var err error
+	var licenseChan *kotsv1beta1.Channel
+	if app.ChannelID == "" {
+		backfillID := kotsutil.GetBackfillChannelIDFromLicense(license)
+		if err := kotsStore.SetAppChannelID(app.ID, backfillID); err != nil {
+			return nil, errors.Wrap(err, "failed to backfill channel id")
+		}
+		if licenseChan, err = kotsutil.FindChannelInLicense(backfillID, license); err != nil {
+			return nil, errors.Wrap(err, "failed to find channel in license")
```
Comment (divolgin) [inline pkg/update/update.go:40]: nit: same error reused in one function

File: pkg/update/update.go [hunk 4]
```diff
@@ -88,8 +104,8 @@ func GetAvailableAirgapUpdates(app *apptypes.App, license *kotsv1beta1.License)
 		if airgap.Spec.AppSlug != license.Spec.AppSlug {
 			return nil
 		}
-		if airgap.Spec.ChannelID != license.Spec.ChannelID {
-			return nil
+		if _, err = kotsutil.FindChannelInLicense(airgap.Spec.ChannelID, license); err != nil {
+			return nil // skip airgap updates that are not for the current channel, preserving previous behavior
```
Comment (divolgin) [inline pkg/update/update.go:108]: would be good to log something non-threatening here on info level to help troubleshooting cases when updates are not showing up.

**Comment by pandemicsyn** [comment]
The approach I took was basically find all occurrences where we using license.spec.ChannelID or license.spec.IsSemverRequired and update those areas. In most cases we now use `app.channel_id` and then use it or the associated channels IsSemverRequired setting.

For dealing with the backfill of `app.channel_id` column , i don't proactively do it - and instead only do it when we go to actually use it. If we see app.channel_id is empty we'll try to back fill it.

**Comment by divolgin** [comment]
This LGTM.  One note about backwards compatibility.  Right now I can run `kots install app-slug/some-channel` and then upload a license from any channel. This would not fail with existing versions of kots, and give me a release from the channel in the license, not the one specified on command line.

With this change, single channel license install will start failing. I think we'll need to allow them to proceed. We can call out this change in behavior in release notes, which is probably acceptable.

**Comment by pandemicsyn** [comment]
> With this change, single channel license install will start failing. I think we'll need to allow them to proceed. We can call out this change in behavior in release notes, which is probably acceptable.

@divolgin Let them proceed with an install for the requested (not in license) channel, or proceed but switch over to the default channel in the supplied license? Guessing the later - the default channel.

**Comment by pandemicsyn** [comment]
Screenshot 2024-07-23 at 12 54 20 PM

File: cmd/kots/cli/install.go [hunk 5]
```diff
@@ -162,6 +162,28 @@ func InstallCmd() *cobra.Command {
 			}()
 
 			upstream := pull.RewriteUpstream(args[0])
+			preferredChannelSlug, err := extractPreferredChannelSlug(upstream)
+			if err != nil {
+				return errors.Wrap(err, "failed to extract preferred channel slug")
+			}
+
+			// fetch the latest version of the license to verify channel access if online install
+			if license != nil {
```
Comment (sgalsaleh) [inline cmd/kots/cli/install.go:171]: the install command is already long and complex, what do you think about moving this logic be to a separate package?

File: cmd/kots/cli/install.go [hunk 6]
```diff
@@ -162,6 +162,28 @@ func InstallCmd() *cobra.Command {
...
+				if isAirgap && haveMultiChannelLicense(license) && !slugInLicenseChannels(preferredChannelSlug, license) {
+					return errors.New("requested channel not found in supplied license")
+				}
+				log.ActionWithSpinner("Checking for license update")
+				updatedLicense, err := replicatedapp.GetLatestLicense(license)
```
Comment (sgalsaleh) [inline cmd/kots/cli/install.go:176]: wouldn't this try to get the latest license from replicated app even in airgap mode?

File: cmd/kots/cli/install.go [hunk 7]
```diff
@@ -162,6 +162,28 @@ func InstallCmd() *cobra.Command {
...
+				if err != nil {
+					log.FinishSpinnerWithError()
+					return errors.Wrap(err, "failed to get latest license")
+				}
+				log.FinishSpinner()
+				if haveMultiChannelLicense(updatedLicense.License) && !slugInLicenseChannels(preferredChannelSlug, updatedLicense.License) {
```
Comment (sgalsaleh) [inline cmd/kots/cli/install.go:182]: how about combining both `haveMultiChannelLicense` and `slugInLicenseChannels` into a single `canInstallFromChannel` function?

File: cmd/kots/cli/install.go [hunk 8]
```diff
@@ -1076,3 +1099,19 @@ func checkPreflightResults(response *handlers.GetPreflightResultResponse, skipPr
...
+func slugInLicenseChannels(slug string, license *kotsv1beta1.License) bool {
+	for _, channel := range license.Spec.Channels {
+		if channel.ChannelSlug == slug {
+			return true
+		}
+	}
+	return false
+}
+
+func haveMultiChannelLicense(license *kotsv1beta1.License) bool {
```
Comment (sgalsaleh) [inline cmd/kots/cli/install.go:1112]:
```suggestion
func isMultiChannelLicense(license *kotsv1beta1.License) bool {
```

Comment (sgalsaleh) [inline cmd/kots/cli/install.go:1112]:
```suggestion
func isMultiChannelLicense(license *kotsv1beta1.License) bool {
```

File: cmd/kots/cli/install.go [hunk 9]
```diff
@@ -1076,3 +1099,19 @@ func checkPreflightResults(response *handlers.GetPreflightResultResponse, skipPr
...
+	if license == nil {
+		return false
+	}
+	return len(license.Spec.Channels) > 0
```
Comment (sgalsaleh) [inline cmd/kots/cli/install.go:1116]: mind adding a comment here that the existence of the `channels` field in the license is what determines if the license is multi channel or not? took me a second to realize why this is `> 0` and not `> 1`.

File: pkg/handlers/license.go [hunk 10]
```diff
@@ -334,7 +334,16 @@ func (h *Handler) UploadNewLicense(w http.ResponseWriter, r *http.Request) {
 	desiredAppName := strings.Replace(verifiedLicense.Spec.AppSlug, "-", " ", 0)
 	upstreamURI := fmt.Sprintf("replicated://%s", verifiedLicense.Spec.AppSlug)
 
-	a, err := store.GetStore().CreateApp(desiredAppName, upstreamURI, licenseString, verifiedLicense.Spec.IsAirgapSupported, installationParams.SkipImagePush, installationParams.RegistryIsReadOnly)
+	// verify that requested channel slug exists in the license
+	matchedChannelID, err := kotsutil.FindChannelIDInLicense(installationParams.RequestedChannelSlug, verifiedLicense)
+	if err != nil {
+		logger.Error(err)
+		uploadLicenseResponse.Error = "Your current license does not grant access to the channel you requested. Please generate a support bundle and contact support for assistance."
```
Comment (sgalsaleh) [inline pkg/handlers/license.go:341]: should this error just be returned by `FindChannelIDInLicense` instead?

File: pkg/handlers/upgrade_service.go [hunk 11]
```diff
@@ -75,6 +75,16 @@ func (h *Handler) StartUpgradeService(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if foundApp.IsAirgap && foundApp.ChannelID != request.ChannelID {
+		// The update is deployable, so we should update the apps stored channelID similar to install
+		// time when channelID is set via the requested-channel-id in the configmap.
+		if err := store.GetStore().SetAppChannelID(foundApp.ID, request.ChannelID); err != nil {
```
Comment (sgalsaleh) [inline pkg/handlers/upgrade_service.go:81]: setting the app channel id should happen [here](https://github.com/replicatedhq/kots/blob/main/pkg/operator/operator.go#L1024) instead, as that's when the update gets deployed, not here. that's because the user might decide to abort the upgrade and not actually deploy this update.

File: pkg/handlers/license.go (hunk 10 above)
Comment (pandemicsyn) [reply pkg/handlers/license.go:341]: I'm not sure, originally i was just returning the terser error that FindChannelIDInLicense returned but Brian word smith'd this for when this bubbles up in the UI. If we're cool with this more verbose one pop'ing up in other places (logs, cli, etc) i can definitely return it directly.

File: cmd/kots/cli/install.go (hunk 6 above)
Comment (pandemicsyn) [reply cmd/kots/cli/install.go:176]: yep, totally right

**Comment by pandemicsyn** [comment]
@sgalsaleh was able to update upstream.Fetch and friends and propagate the app channel id like we chatted about. I didn't have to go far - i was able to basically pull it from PendingApp or App in all the main cases.  Its all in this commit: https://github.com/replicatedhq/kots/pull/4767/commits/a5e4bd9f79e9611357969efe0ba0be7da5f65581 incase thats a bit easier. 

I think I'll want/need to update some of these tests so I'll probably steal a bit more time on your calendar to pepper you with additional questions about those. To get the `tests/pull/*` passing for now I just updated the testcases.yaml's pull options with the single app ChannelID we used everywhere.

File: pkg/multichannel/multichannel.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// VerifyAndUpdateLicense will update (if not airgapped), verify that the request channel slug is present, and return the possibly updated license.
+// Note that this is a noop if the license passed in is nil.
+func VerifyAndUpdateLicense(log *logger.CLILogger, license *kotsv1beta1.License, preferredChannelSlug string, isAirgap bool) (*kotsv1beta1.License, error) {
+	if license == nil {
+		return nil, nil
+	}
+	if isAirgap {
+		if isMultiChannelLicense(license) && !isSlugInLicenseChannels(preferredChannelSlug, license) {
```
Comment (sgalsaleh) [inline pkg/multichannel/multichannel.go:43]:
```suggestion
		if !canInstallFromChannel(preferredChannelSlug, license) {
```

File: pkg/store/kotsstore/downstream_store.go [hunk 13]
```diff
@@ -673,8 +686,21 @@ func (s *KOTSStore) AddDownstreamVersionsDetails(appID string, clusterID string,
...
+			}
+		} else {
+			if licenseChan, err = kotsutil.FindChannelInLicense(foundChannelID, license); err != nil {
+				errors.Wrap(err, "failed to find channel in license")
+			}
+		}
+
 		for _, v := range versions {
-			v.IsDeployable, v.NonDeployableCause = isAppVersionDeployable(v, allVersions, license.Spec.IsSemverRequired)
+			v.IsDeployable, v.NonDeployableCause = isAppVersionDeployable(v, allVersions, licenseChan.IsSemverRequired)
```
Comment (sgalsaleh) [inline pkg/store/kotsstore/downstream_store.go:703]: wouldn't `licenseChan` be `nil` if `foundChannelID` is empty?

File: pkg/store/kotsstore/downstream_store.go [hunk 14]
```diff
@@ -409,7 +409,20 @@ func (s *KOTSStore) GetDownstreamVersions(appID string, clusterID string, downlo
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to get app license")
 	}
-	downstreamtypes.SortDownstreamVersions(result.AllVersions, license.Spec.IsSemverRequired)
+
+	foundChannelID, err := s.GetAppChannelID(appID)
+	var licenseChan *kotsv1beta1.Channel
+	if foundChannelID == "" {
```
Comment (sgalsaleh) [inline pkg/store/kotsstore/downstream_store.go:415]: i see this code duplicated in numerous locations. can it be moved to a shared function? something like `BackfillAppChannelID`.

File: pkg/update/required.go [hunk 15]
```diff
@@ -64,9 +64,14 @@ func getRequiredAirgapUpdates(airgap *kotsv1beta1.Airgap, license *kotsv1beta1.L
 		for _, appVersion := range installedVersions {
 			requiredSemver, requiredSemverErr := semver.ParseTolerant(requiredRelease.VersionLabel)
 
+			licenseChan, err := kotsutil.FindChannelInLicense(appVersion.ChannelID, license)
```
Comment (sgalsaleh) [inline pkg/update/required.go:67]: this should use the app channel id, not the app version channel id. that's because the app version one would be old. we should always use the most recent channel id, that previously came from the license, but with this PR, it should come from the app like everywhere else.

File: pkg/update/required_test.go [hunk 16]
```diff
@@ -12,6 +12,29 @@ import (
 
 func Test_getRequiredAirgapUpdates(t *testing.T) {
```
Comment (sgalsaleh) [inline pkg/update/required_test.go:13]: this test should be updated given my comment above about using the channel id from the app

File: pkg/update/update.go [hunk 17]
```diff
@@ -29,7 +31,21 @@ func InitAvailableUpdatesDir() error {
 }
 
 func GetAvailableUpdates(kotsStore storepkg.Store, app *apptypes.App, license *kotsv1beta1.License) ([]types.AvailableUpdate, error) {
-	updateCursor, err := kotsStore.GetCurrentUpdateCursor(app.ID, license.Spec.ChannelID)
+	var err error
+	var licenseChan *kotsv1beta1.Channel
+	if app.ChannelID == "" {
```
Comment (sgalsaleh) [inline pkg/update/update.go:36]: would using `GetOrBackfillLicenseChannel` suffice here?

File: pkg/updatechecker/updatechecker.go [hunk 18]
```diff
@@ -226,7 +228,20 @@ func checkForKotsAppUpdates(opts types.CheckForUpdatesOpts, finishedChan chan<-
 		return nil, errors.Wrap(err, "failed to get app")
 	}
 
-	updateCursor, err := store.GetCurrentUpdateCursor(a.ID, latestLicense.Spec.ChannelID)
+	var licenseChan *kotsv1beta1.Channel
+	if a.ChannelID == "" {
```
Comment (sgalsaleh) [inline pkg/updatechecker/updatechecker.go:232]: ditto: wouldn't using `GetOrBackfillLicenseChannel` suffice here?

File: pkg/handlers/update_checker_spec.go [hunk 19]
```diff
@@ -56,8 +57,27 @@ func (h *Handler) SetAutomaticUpdatesConfig(w http.ResponseWriter, r *http.Reque
 		return
 	}
 
+	var licenseChan *kotsv1beta1.Channel
+	if foundApp.ChannelID == "" {
```
Comment (sgalsaleh) [inline pkg/handlers/update_checker_spec.go:61]: ditto: wouldn't using `GetOrBackfillLicenseChannel` suffice here?

File: pkg/kotsutil/kots.go [hunk 20]
```diff
@@ -1598,3 +1600,62 @@ func GetECVersionFromAirgapBundle(airgapBundle string) (string, error) {
...
+	for _, channel := range license.Spec.Channels {
+		if channel.ChannelID == channelID {
+			return &channel, nil
+		}
+	}
+
+	return nil, errors.New("channel not found in multi channel format license")
+}
+
+func GetBackfillChannelIDFromLicense(license *kotsv1beta1.License) string {
```
Comment (sgalsaleh) [inline pkg/kotsutil/kots.go:1652]: how about renaming this to
```suggestion
func GetDefaultChannelIDFromLicense(license *kotsv1beta1.License) string {
```
frankly, this should be part of the license type in kots kinds and named `GetDefaultChannelID` or `DefaultChannelID`, but it's fine for now.

File: pkg/kotsutil/kots.go [hunk 21]
```diff
@@ -1598,3 +1600,62 @@ func GetECVersionFromAirgapBundle(airgapBundle string) (string, error) {
...
+	} else { // this is an install from before the channel slug was added to the configmap
+		logger.Debug("requested channel slug not found in configmap, using top level channel id from license")
+		matchedChannelID = license.Spec.ChannelID
+	}
+	return matchedChannelID, nil
+}
+
+func FindChannelInLicense(channelID string, license *kotsv1beta1.License) (*kotsv1beta1.Channel, error) {
+	if channelID == "" || len(license.Spec.Channels) == 0 {
+		if license.Spec.ChannelID != channelID {
```
Comment (sgalsaleh) [inline pkg/kotsutil/kots.go:1631]: wouldn't this always be false if the channel id is empty?

File: pkg/multichannel/multichannel.go [hunk 22]
```diff
@@ -0,0 +1,59 @@
+package multichannel
```
Comment (sgalsaleh) [inline pkg/multichannel/multichannel.go:1]: any reason this is not in the existing `license` package?

File: pkg/pull/pull.go [hunk 23]
```diff
@@ -236,7 +238,7 @@ func Pull(upstreamURI string, pullOptions PullOptions) (string, error) {
 			logger.Infof("Expecting to install version %s but airgap bundle version is %s.", fetchOptions.AppVersionLabel, airgap.Spec.VersionLabel)
 		}
 
-		if fetchOptions.License.Spec.ChannelID != airgap.Spec.ChannelID {
+		if _, err = kotsutil.FindChannelInLicense(airgap.Spec.ChannelID, fetchOptions.License); err != nil {
 			return "", util.ActionableError{
 				NoRetry: true, // if this is airgap upload, make sure to free up tmp space
 				Message: fmt.Sprintf("License (%s) and airgap bundle (%s) channels do not match.", fetchOptions.License.Spec.ChannelName, airgap.Spec.ChannelName),
```
Comment (sgalsaleh) [inline pkg/pull/pull.go:244]: this error message will be confusing for multi-channel licenses as it's using the top-level channel name.

File: migrations/tables/app.yaml [hunk 24]
```diff
@@ -85,3 +85,5 @@ spec:
         default: 0
         constraints:
           notNull: true
+      - name: channel_id
```
Comment (sgalsaleh) [inline migrations/tables/app.yaml:88]: how about renaming this to
```suggestion
      - name: selected_channel_id
```
this way it's clearer what this column represents. preferably rename it in the app type/struct as well.

File: pkg/multichannel/multichannel.go (hunk 22 above)
Comment (pandemicsyn) [reply pkg/multichannel/multichannel.go:1]: ah - originally had more in here and ended up with some circular imports when i had it under the `license` package but its pretty small now and the only other pkg import is replicatedapp.

File: pkg/update/update.go (hunk 17 above)
Comment (pandemicsyn) [reply pkg/update/update.go:36]: Yep i can switch over to using that. On these we already know app.ChannelID is empty, so calling GetOrBackfill generates an extra call to the db to see if its present. That's really the only reason these where backfilling directly instead.

Comment (sgalsaleh) [reply pkg/update/update.go:36]: > On these we already know app.ChannelID is empty

how so? wouldn't it be possible that it go backfilled at some point?

Comment (sgalsaleh) [reply pkg/update/update.go:36]: also, the extra db call would only happen once to backfill the channel id, no?

**Comment by sgalsaleh** [comment]
@pandemicsyn we should add an e2e test.

**Comment by sgalsaleh** [comment]
I might've missed it, but where does KOTS send the user-selected channel ID to replicated.app?

File: pkg/update/update.go (hunk 17 above)
Comment (pandemicsyn) [reply pkg/update/update.go:36]: The `GetOrBackfillLicenseChannel` method queries the app table for the channel id using the app id provided. Then either just returns that channel id if present or backfill's if we didn't have one (and then returns the backfilled id). 

But in some of these call locations we've just recently called `store.GetApp()`. So `app.ChannelID` was already fresh. On those instead of calling `GetOrBackfillLicenseChannel` i was just backfilling to skip having to hit the table again just to find out that channel id is empty.

File: pkg/update/required.go (hunk 15 above)
Comment (pandemicsyn) [reply pkg/update/required.go]: Update in Commit: https://github.com/replicatedhq/kots/pull/4767/commits/45ababaa752aae2fe08ccd71a64a7748d056a4a3

File: migrations/tables/app.yaml (hunk 24 above)
Comment (pandemicsyn) [reply migrations/tables/app.yaml]: Renamed in: https://github.com/replicatedhq/kots/pull/4767/commits/c39c4553a3b154a1bdfe09fdfea78df9464d3fd1

File: pkg/multichannel/multichannel.go (hunk 22 above)
Comment (pandemicsyn) [reply pkg/multichannel/multichannel.go]: relocated in:  https://github.com/replicatedhq/kots/pull/4767/commits/61ea5d54becda901b8e5e86a0a67f8569d7b5f45

File: pkg/pull/pull.go (hunk 23 above)
Comment (pandemicsyn) [reply pkg/pull/pull.go]: fixed this up in this catch all commit: https://github.com/replicatedhq/kots/pull/4767/commits/f2624c8e7d5bb2a98151a79980d78de23806130f

**Comment by pandemicsyn** [comment]
This will need https://github.com/replicatedhq/vandoor/pull/5889 but this commit should cover all the endpoints that now accept it on the replicated app side: https://github.com/replicatedhq/kots/pull/4767/commits/f38716c447c2a52fc73fa9cb5605707cde55bacd

I did skip updating [`downloadUpgradeBinary`](https://github.com/replicatedhq/kots/blob/5670906832ce9c0087e37d425c60317a80872f08/pkg/embeddedcluster/upgrade.go#L220) call where embedded cluster calls out to get `/clusterconfig/artifact/operator` , because I wasn't sure if i should actually be updating this one.

**Comment by pandemicsyn** [comment]
@sgalsaleh @divolgin i centralized the post license sync update of `selected_channel_id` in https://github.com/replicatedhq/kots/pull/4767/commits/fb6e5519f563c997919222602719a6b2cc4a89dc. I did it in the `UpdateAppLicense` because that seemed like to safest place to do it for consistencies sake.  Let me know if miss-read this, or if theres a better spot though, happy to switch it.

Also nuked that stray selected channel id update in reconcileDeployment  in this commit: https://github.com/replicatedhq/kots/pull/4767/commits/ab0f4d7786d91a4ab76ffa3c37e9d4b6109a5303

**Comment by pandemicsyn** [comment]
Ended up centralizing the backfilling, I moved it into store.GetApp() ...tbh...not sure why I didn't stick it there to begin with. 

Also, heres a loom of license channel change working (going from "beta" with a v1.0.2 to "unstable" which was up to v1.0.3): https://www.loom.com/share/5b89a5a3423341f881c7510a2149b2fb?sid=4a072c59-b965-47b0-bb2d-7c5187b511c3

File: pkg/tests/renderdir/renderdir_test.go [hunk 25]
```diff
@@ -73,7 +73,7 @@ func TestKotsRenderDir(t *testing.T) {
 			Name:             spec.Name,
 			RenderDirOptions: spec.RenderDirOptions,
 		}
-
+		test.RenderDirOptions.App.SelectedChannelID = "1vusIYZLAVxMG6q760OJmRKj5i5"
```
Comment (sgalsaleh) [inline pkg/tests/renderdir/renderdir_test.go:76]: why is this needed if you're already passing it in the render options?

File: pkg/update/required_test.go [hunk 26]
```diff
@@ -164,8 +186,68 @@ func Test_getRequiredAirgapUpdates(t *testing.T) {
...
+							ChannelName:      "Stable Channel",
+							ChannelSlug:      "stable-channel",
+							IsDefault:        true,
+							IsSemverRequired: true,
+						},
+						{
+							ChannelID:        "different-channel",
+							ChannelName:      "Different Channel",
+							ChannelSlug:      "different-channel",
+							IsDefault:        true,
```
Comment (sgalsaleh) [inline pkg/update/required_test.go:239]: i see two default channels in here. is that expected?

File: pkg/update/required_test.go [hunk 27]
```diff
@@ -164,8 +186,68 @@ func Test_getRequiredAirgapUpdates(t *testing.T) {
...
-				Spec: kotsv1beta1.LicenseSpec{},
+				Spec: kotsv1beta1.LicenseSpec{
+					ChannelID:   "stable-channel", // intentionally fully avoiding the default channel
+					ChannelName: "Stable Channel",
+					Channels: []kotsv1beta1.Channel{
+						{
+							ChannelID:        "stable-channel",
+							ChannelName:      "Stable Channel",
+							ChannelSlug:      "stable-channel",
+							IsDefault:        true,
```
Comment (sgalsaleh) [inline pkg/update/required_test.go:232]: should this be `false` given your comment above?

File: pkg/update/update_test.go [hunk 28]
```diff
@@ -150,6 +153,44 @@ func TestGetAvailableUpdates(t *testing.T) {
...
+					},
+				},
+			},
+			channelReleases: []upstream.ChannelRelease{},
+			setup: func(t *testing.T, args args, licenseEndpoint string) {
+				t.Setenv("USE_MOCK_REPORTING", "1")
+				args.license.Spec.Endpoint = licenseEndpoint
+				mockStore.EXPECT().GetCurrentUpdateCursor(args.app.ID, args.license.Spec.Channels[1].ChannelID).Return("1", nil)
+			},
+			want:    []types.AvailableUpdate{},
```
Comment (sgalsaleh) [inline pkg/update/update_test.go:191]: can you please update the `newMockServerWithReleases` below to respect `selectedChannelID` query param similar to replicated app, and then validate that the available updates are returned for the correct channel?

File: pkg/updatechecker/updatechecker.go [hunk 29]
```diff
@@ -226,7 +228,13 @@ func checkForKotsAppUpdates(opts types.CheckForUpdatesOpts, finishedChan chan<-
 		return nil, errors.Wrap(err, "failed to get app")
 	}
 
-	updateCursor, err := store.GetCurrentUpdateCursor(a.ID, latestLicense.Spec.ChannelID)
+	var licenseChan *kotsv1beta1.Channel
+	licenseChan, err = kotsutil.FindChannelInLicense(a.SelectedChannelID, latestLicense)
```
Comment (sgalsaleh) [inline pkg/updatechecker/updatechecker.go:232]: ditto, why not just do this:
```suggestion
	licenseChan, err := kotsutil.FindChannelInLicense(a.SelectedChannelID, latestLicense)
```

File: pkg/update/update.go [hunk 30]
```diff
@@ -29,7 +31,15 @@ func InitAvailableUpdatesDir() error {
 }
 
 func GetAvailableUpdates(kotsStore storepkg.Store, app *apptypes.App, license *kotsv1beta1.License) ([]types.AvailableUpdate, error) {
-	updateCursor, err := kotsStore.GetCurrentUpdateCursor(app.ID, license.Spec.ChannelID)
+	var err error
+	var licenseChan *kotsv1beta1.Channel
+
+	licenseChan, err = kotsutil.FindChannelInLicense(app.SelectedChannelID, license)
```
[the remark this answers is not in the document]
Comment (pandemicsyn) [reply pkg/update/update.go:37]: Yep this is totally fine, either i just over looked it or it in a previous verison I had an `if` in the mix.

File: pkg/tests/renderdir/renderdir_test.go (hunk 25 above)
Comment (pandemicsyn) [reply pkg/tests/renderdir/renderdir_test.go:76]: Its not - honestly don't even recall when/why i added this - sorry bout that.

File: pkg/update/update_test.go (hunk 28 above)
Comment (pandemicsyn) [reply pkg/update/update_test.go:191]: 👍 yep - also updated the multi chan test to actually do some work.

**Comment by pandemicsyn** [comment]
@sgalsaleh alrighty - pulled in the e2e test changes as well (https://github.com/replicatedhq/kots/pull/4767/commits/3961a04f1e5dc4f998e25ef96acfea43983f94a2)  , batched the other pr feedback into a single commit this time since it was pretty small (https://github.com/replicatedhq/kots/pull/4767/commits/3cd0a798374c5801b71b33d0b34eebad2ca5a489)

**Review by sgalsaleh — APPROVED** [review]
